src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -459,8 +481,459 @@ public void WithAnalyzerConfigDocumentTextLoader()
+
+            var newSolution = solution.WithProjectParseOptions(projectId, options);
+            Assert.Same(options, newSolution.GetProject(projectId)!.ParseOptions);
+
+            Assert.Same(newSolution, newSolution.WithProjectParseOptions(projectId, options));
+
+            Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectParseOptions(null!, options));
+            Assert.Throws<ArgumentNullException>("options", () => solution.WithProjectParseOptions(projectId, null!));
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectParseOptions(ProjectId.CreateNewId(), options));
+        }
+
+        [Fact]
+        public void WithProjectReferences()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var projectId = solution.Projects.Single().Id;
+
+            var projectId2 = ProjectId.CreateNewId();
+            solution = solution.AddProject(projectId2, "proj2", "proj2.dll", LanguageNames.CSharp);
+            var projectRef = new ProjectReference(projectId2);
+
+            SolutionTestHelpers.TestListProperty(solution,
+                (old, value) => old.WithProjectReferences(projectId, value),
+                opt => opt.GetProject(projectId)!.AllProjectReferences,
+                projectRef,
+                allowDuplicates: false);
+
+            var projectRefs = (IEnumerable<ProjectReference>)ImmutableArray.Create(
+                new ProjectReference(projectId2),
+                new ProjectReference(projectId2, ImmutableArray.Create("alias")),
+                new ProjectReference(projectId2, embedInteropTypes: true));
+
+            var solution2 = solution.WithProjectReferences(projectId, projectRefs);
+            Assert.Same(projectRefs, solution2.GetProject(projectId)!.AllProjectReferences);
+
+            Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectReferences(null!, new[] { projectRef }));
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectReferences(ProjectId.CreateNewId(), new[] { projectRef }));
+        }
+
+        [Fact]
+        [WorkItem(42406, "https://github.com/dotnet/roslyn/issues/42406")]
+        public void WithProjectReferences_ProjectNotInSolution()
+        {
+            var solution = CreateSolutionWithProjectAndDocuments();
+            var projectId = solution.Projects.Single().Id;
+            var externalProjectRef = new ProjectReference(ProjectId.CreateNewId());
+
+            var projectRefs = (IEnumerable<ProjectReference>)ImmutableArray.Create(externalProjectRef);
+            var newSolution1 = solution.WithProjectReferences(projectId, projectRefs);
+            Assert.Same(projectRefs, newSolution1.GetProject(projectId)!.AllProjectReferences);
+
+            // project reference is not included:
+            Assert.Empty(newSolution1.GetProject(projectId)!.ProjectReferences);
```

Comment:
Do we also have tests confirming the dependency graph is doing the right things? I would have thought so but if you observed something was busted maybe not? Or we were testing those at the dependency graph level and not the outright Solution API level?

Reply:
I don't. Filed a follow up issue: https://github.com/dotnet/roslyn/issues/42450

Reply:
There's already a huge pile of tests in https://github.com/dotnet/roslyn/blob/master/src/Workspaces/CoreTest/SolutionTests/ProjectDependencyGraphTests.cs so this one might be covered or might be easy to add.